### DISCUSSION TITLE
State representation: joint position and angle, or separate?

### DISCUSSION THREAD

**Discussion by knyazer** [DelftMercurians/cotix#36] [Ideas]
So, there are two chairs: one has joint representation on it, the other one has separate into two representations. We will have to choose a chair....

Currently we are sitting on a separate position/angle representation, but I am contemplating changing seats.

### Why to choose joint representation:
* easier to pass around
* two less variables in the ```Body```
* easier to apply velocities
* probably considered a standard? [wiki, under equations of motions](https://en.wikipedia.org/wiki/Multibody_system)
### Why not to choose joint representation:
* harder to comprehend: ```.state[2]``` and ```.angle``` (obviously second one is better, but we will have to use first one)
* harder(?) to do complicated manipulations. Like, imagine you are trying to figure out distance between two bodies, and you will have to do something like ```jnp.linalg.norm(a.state[:2] - b.state[:2])``` :vomiting_face: 

WDYT?

**Comment by knyazer** [comment]
@mikhail-vlasenko @ivrolan (cannot ping anybody else lol)

**Comment by mikhail-vlasenko** [comment]
i think the current thing is fine, the `jnp.linalg.norm(a.state[:2] - b.state[:2])` really sucks, while I don't think `easier to apply velocities` is much of a pro, as we probably will just have a single function to do that. don't really have a strong preference though

**Comment by ivrolan** [comment]
I don't think having a state vector like that could be a problem, it is widely used. However, we can add some getters to the class instead of accessing directly the members variables for convenience. Something like:
    
    a.get_position() # -> returns a.state[:2]
    a.get_angle() # -> returns a.state[2]

**Reply by knyazer** [reply]
hmm, interesting. The fact that something is widely used does not necessarily mean that it is a good decision: 20 years ago there was no 'code quality', so all the programmers were using math-y notation. But I also don't say that it is bad. It is more like a bit good, not super good.

Having getters is another questionable practice: while, since we do have an invariant checker, getter effectively guarantees that the value you get is 'correct' to some degree (aka not nan), it is still not commonly considered good. Again, this is the same point as with math-y notation :)

Also we cannot really prohibit from not-using direct attribute access, though we could try by making fields with underscores, that would signify that you should use them only if you understand what you are doing. Actually, it starts looking better for me now...

Overwriting \_\_setattr\_\_ and \_\_getattr\_\_ seems like a way to go, but compatibility with jax might be hurt. Or might not. Should check it

Anyway, a good suggestion! Thanks!

**Reply by knyazer** [reply]
Yeah, since I am on the side of strictness if the expressivity is not hurt, I think I am gonna take a look at forcing you not to use attributes directly at some point. There is of course ```eqx.tree_at```, but if you use it you probably know what you are doing. 

Beside the point, would be nice to have a python analyzer, that would detect when you do something illegal with JAX statically/tracing-time. Like, for example, there is a lot of syntax that is not allowed:
```python
class A(eqx.Module):
     def do_smth():
          ...

A.do_smth() # by definition does nothing, and will be removed by XLA compiler. We want to tell the user that he is doing something wrong? The problem becomes to detect this outside jitted regions, or whatever.
